### PR TITLE
Kostentraegerdatei fetching

### PR DESCRIPTION
**Pull request by westnordost**
Necessary changes after testing kostenträgerdatei parsing with real data:
- more lenient: If one segment of a message fails parsing, only that segment is skipped. If one message fails parsing, only that message is skipped. But return these warnings in the parse result.
- added/removed fields in parse result

Current parse result as a JSON:
https://gist.github.com/westnordost/92bb15716928def837685d20635b79db

### CODE REVIEW

File: src/kostentraeger/transformer.ts [hunk 1]
```diff
@@ -102,6 +125,10 @@ function createReceiptTransmissionMethods(uemList: UEM[], dfuList: DFU[]): Recei
         ftam = dfuList.find((dfu) => dfu.dfuProtokollSchluessel == "016")?.address
     }
 
+    if(!paperReceipt && !machineReadablePaperReceipt && !email && !ftam) {
+        throw new Error(`Expected that institution accepts either paper receipts or sent via email or ftam`)
```
Comment (westnordost) [inline src/kostentraeger/transformer.ts:129]: FYI @michaelkamphausen :  Most accept email, but there is a handful of institutions that only have a FTAM address.

**Comment by westnordost** [comment]
The file size of that JSON is around 12MB. The combined size of all `.ke*` Kostenträger files is 1MB.

**Comment by westnordost** [comment]
Current warnings are:
```
LK06Q121.ke0
	message 18 - skipped invalid VKG 3: Unknown LeistungserbringergruppeSchluessel "3"
	message 18 - skipped invalid VKG 4: Unknown LeistungserbringergruppeSchluessel "3"
	message 18 - skipped invalid VKG 5: Unknown LeistungserbringergruppeSchluessel "3"
	message 18 - skipped invalid VKG 6: Unknown LeistungserbringergruppeSchluessel "3"
	
AO06Q221.ke1
	message 185 - skipped invalid UEM 1: Refers to a non-existing DFU
	message 186 - skipped invalid UEM 1: Refers to a non-existing DFU
	
AO05Q221.ke1
	message 106 - skipped invalid DFU 1: Unknown DFUProtokollSchluessel "011"
	message 106 - skipped invalid DFU 2: Unknown DFUProtokollSchluessel "011"
	message 106 - skipped invalid DFU 3: Unknown DFUProtokollSchluessel "011"
	message 136 - skipped invalid DFU 1: Unknown DFUProtokollSchluessel "011"
	message 153 - skipped invalid DFU 1: Unknown DFUProtokollSchluessel "011"
	message 153 - skipped invalid DFU 2: Unknown DFUProtokollSchluessel "004"
	message 153 - skipped invalid DFU 3: Unknown DFUProtokollSchluessel "011"
	message 202 - skipped invalid UEM 1: Refers to a non-existing DFU
	
BK05Q221.ke1
	message 303 - skipped invalid VKG 11: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 12: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 13: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 14: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 15: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 16: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 17: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 18: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 20: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 21: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 22: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 23: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 24: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 25: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 26: Links to data acceptance office that does not accept data
	message 303 - skipped invalid VKG 27: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 11: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 12: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 13: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 14: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 15: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 16: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 17: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 18: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 20: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 21: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 22: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 23: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 24: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 25: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 26: Links to data acceptance office that does not accept data
	message 304 - skipped invalid VKG 27: Links to data acceptance office that does not accept data

BN06Q318.ke0
	message 29 - skipped invalid DFU 1: Unknown DFUProtokollSchluessel "002"
	message 29 - skipped invalid DFU 2: Unknown DFUProtokollSchluessel "011"
	message 29 - skipped invalid DFU 3: Unknown DFUProtokollSchluessel "030"
	message 29 - skipped invalid DFU 4: Unknown DFUProtokollSchluessel "050"

BN050321.ke0
	message 28 - skipped invalid DFU 1: Unknown DFUProtokollSchluessel "002"
	message 28 - skipped invalid DFU 2: Unknown DFUProtokollSchluessel "011"
	message 28 - skipped invalid DFU 3: Unknown DFUProtokollSchluessel "030"
	message 28 - skipped invalid DFU 4: Unknown DFUProtokollSchluessel "050"
```

All of these should be harmless. Regarding those for BK05Q221.ke1, there might be a misunderstanding. Will need to contact the institution about it.

**Comment by michaelkamphausen** [comment]
This is great! I like how forgiving the parser is and issues warnings and it's very nice to see the parsed data.

**Review by michaelkamphausen — COMMENTED** [review]
> The file size of that JSON is around 12MB. The combined size of all `.ke*` Kostenträger files is 1MB.

12MB with or without indentation and newline whitespace?
Turns out the Kostenträger file format is indeed reasonably efficient for storage. 🤔

File: src/kostentraeger/fetcher.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    return await Promise.all(kostentraegerFileUrls.map(async (url) => await fetchKostentraegerFile(url)))
+}
+
+async function fetchKostentraegerFile(url: string): Promise<InstitutionList> {
+    const response = await fetch(url)
+    /* Kostenträger files are encoded in iso-8859-1 and not in UTF-8, so we cannot
+       just call response.text()! */
+    // TODO but TextDecoder is not available in JSDOM test environment!
+    //const decoder = new TextDecoder("iso-8859-1")
+    //const text = decoder.decode(await response.arrayBuffer())
```
[the remark this answers is not in the document]
Comment (westnordost) [reply src/kostentraeger/fetcher.ts:38]: Well, `TextDecoder` is available in all modern browsers I read. Problem is, that Jest is neither browser nor node, bug apparently (didn't read that deep into it yet) runs in an environment called JSDOM.

**Comment by westnordost** [comment]
> 12MB with or without indentation and newline whitespace?

With pretty print. Minified, it is a bit over 5MB. I was thinking, it could work like this:

Task fetches the Kostenträger files, does not parse them, but puts them into a JSON like this,
```js
{
  "AO06Q221.ke1": "the content of that file",
  "BK06Q221.ke1": "the content of that other file",
  // etc...
}
```

Then, either puts them into local storage or puts it as one file into the distribution. The parser can then parse that JSON file into memory.

File: src/kostentraeger/fetcher.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (michaelkamphausen) [reply src/kostentraeger/fetcher.ts:38]: Give it a try. After adding the import statement and uncommenting the two lines, I ran the tests and they succeeded without error.
I was wondering how that exact import statement would behave when the transpiled code is executed in a browser: maybe it just works or maybe it doesn't and therefore needs to be conditional.

**Comment by michaelkamphausen** [comment]
I agree on using the Kostenträger files for storage and parsing them when needed.

File: src/kostentraeger/fetcher.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (westnordost) [reply src/kostentraeger/fetcher.ts]: You are right, it works! Thank you!

File: src/kostentraeger/index.ts [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+     *        institutions given in the Kostenträger files have a validity date range.
+     */
+    find(
+        paperDataType: PaperDataType,
+        pflegekasseIK: string,
+        leistungsart: Leistungsart,
+        location: CareProviderLocationSchluessel,
+        date: Date = new Date()
+
+    ): KostentraegerFindResult | undefined {
```
Comment (westnordost) [inline src/kostentraeger/index.ts:98]: This is the streamlined interface to query the data in the end, @michaelkamphausen

File: src/kostentraeger/index.ts [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+export type KostentraegerFindResult = {
+    /** Information of the institution with the IK given as pflegekasseIK parameter in the find function */
+    pflegekasse: Institution,
+    /** Information on the instutition which shall be the Kostenträger of this receipt */
+    kostentraeger: Institution,
+    /** Information on the institution for which the receipt shall be encrypted */
+    encryptTo?: Institution,
+    /** Information on the institution to which the receipt shall be sent */
+    sendTo: Institution 
+}
```
Comment (westnordost) [inline src/kostentraeger/index.ts:31]: This is the data that is returned in the end @michaelkamphausen 

I was not sure if you just need the IKs or also other info, so for now, all information about the institution is included.

**Comment by westnordost** [comment]
Did **a lot** proofing to actual data.

- A lot of data could be culled from the data model because it was never specified  (bank account info, many more) or superfluous and/or unreliable (part of UEM, many more)
- Other data could be merged because it was only defined so cumbersome due to limitations of the (EDIFACT) format (links to paper acceptance offices)
- Even other data could be simplified a lot (f.e. bezirkSchluessel)

So finally, the **kostentraeger.min.json is only 886 kB**. There is potential for even more merging, but I'd like to move on now.

File: dist/kostentraeger-warnings.txt [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  IK 101520078 (Mobil Krankenkasse) links to IK 101520078 for data but neither that IK nor an IK it links to accepts data
+  IK 101520078 (Mobil Krankenkasse) links to IK 101520078 for data but neither that IK nor an IK it links to accepts data
+  IK 101520078 (Mobil Krankenkasse) links to IK 101520078 for data but neither that IK nor an IK it links to accepts data
+  IK 101520078 (Mobil Krankenkasse) links to IK 101520078 for data but neither that IK nor an IK it links to accepts data
+  IK 101520078 (Mobil Krankenkasse) links to IK 101520078 for data but neither that IK nor an IK it links to accepts data
+  IK 103524394 (VIACTIV Krankenkasse Ost) links to IK 102193234 for data but neither that IK nor an IK it links to accepts data
+  IK 103524407 (VIACTIV Krankenkasse) links to IK 102193234 for data but neither that IK nor an IK it links to accepts data
+  IK 104526376 (VIACTIV Krankenkasse) links to IK 102193234 for data but neither that IK nor an IK it links to accepts data
+  IK 104529255 (VIACTIV Krankenkasse Ost) links to IK 102193234 for data but neither that IK nor an IK it links to accepts data
```
Comment (westnordost) [inline dist/kostentraeger-warnings.txt:44]: This is  #29

File: tsconfig.json [hunk 6]
```diff
@@ -4,7 +4,7 @@
 
     /* Basic Options */
     // "incremental": true,                         /* Enable incremental compilation */
-    "target": "es5",                                /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
+    "target": "es2015",                             /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
```
Comment (westnordost) [inline tsconfig.json:7]: that was necessary but I don't remember exactly for what. Something to do with proper type inference when iterating something, I think.

Comment (michaelkamphausen) [reply tsconfig.json:7]: I need es6/es2015 for #17 as well. Should be fine as CoopCare WebApp requires es6 anyway.

File: src/kostentraeger/index.ts [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+    let mostCurrentValidityStartDate: Date | undefined
+    for (let i = 0; i < lists.length; i++) {
+        const list = lists[i]
+        if (list.validityStartDate < date) {
+            if (!mostCurrentValidityStartDate || list.validityStartDate > mostCurrentValidityStartDate) {
+                result = list.institutions
+                mostCurrentValidityStartDate = list.validityStartDate
+            }
+        }
+    }
```
Comment (michaelkamphausen) [inline src/kostentraeger/index.ts:172]: I'd suggest two minor simplifications:

```suggestion
    let mostCurrentValidityStartDate = new Date(0) // 1970
    lists.forEach(list => {
        if (list.validityStartDate < date && list.validityStartDate > mostCurrentValidityStartDate) {
            result = list.institutions
            mostCurrentValidityStartDate = list.validityStartDate
        }
    })
```

**Review by michaelkamphausen — APPROVED** [review]

File: src/kostentraeger/index.ts (hunk 4 above)
Comment (michaelkamphausen) [reply src/kostentraeger/index.ts]: Looks good and pretty much complete to me.

**Comment by michaelkamphausen** [comment]
I'm done with code review and excited about the new lookup functionality.

File: src/kostentraeger/index.ts (hunk 7 above)
Comment (westnordost) [reply src/kostentraeger/index.ts:172]: Ah right. The function in its first version returned an `indexOf`, that's why there is still the fori-loop.

**Pull request merged**